### PR TITLE
Events when a connection error occurs and a lot of refactoring

### PR DESCRIPTION
**Pull request by SamuelHaidu**
#### Solved the TODO: "Also create events when things go wrong, like a connection with a node has failed."

Created two methods on Node class: `Node.outbound_node_connection_error()` and `Node.inbound_node_connection_error()`
the two are invoked when occurs a error on `Node.connect_with_node()` or `Node.run()`
The methods receive as a parameter a Exception object.

#### Removed unnecessary comments

Some comments are redundant like email, and author, version (not updated). Everything is in `setup.py` or `__author__` `__version__` module attributes.

#### Docstrings adhering to the PEP 275 convention

Some docstrings is located in non conventional places. Module docstrings is on top. Removed docstrings from class constructor  and put on class top. Some documentations redundancies are removed. Its help the IDEs to find the documentation,

####  Spacings and other thing adhering PEP 8 convention

Add spaces and break lines. Replace `== None ` comparison to `is None`

#### Adds annotation in all methods

Adds python annotations in all methods. Its help the IDEs show autocomplete  and code errors, besides making it more documented.

#### Replace legacy string formatting

Replaced the legacy python 2 `%` , `str.format()` or string concatenation to modern fstrings. Too much more readable.

#### Refactoring methods

`Node.send_to_nodes()` KISS using filter and `self.all_nodes` attribute

`Node.send_to_node()` KISS using addition assignment and `self.all_nodes` attribute

`Node.connect_with_node` A bug was fixed using a loop to check if every inbound nodes has the same host or id. 

##### This has been simplified by
**Before connection**
Check if every connected node (inbound or outbound) has the same host or port.

**After the connection**
Check if every node  (inbound or outbound) has the same ID

Now the checks types are separated.

### CODE REVIEW

**Comment by macsnoeren** [comment]
Wow! I will have a look in the coming weeks. Happy new year and thanks for the contribution!

**Comment by SamuelHaidu** [comment]
Thanks, happy new year!!  I'm preparing more commits hehehe

File: p2pnetwork/node.py [hunk 1]
```diff
@@ -75,114 +65,98 @@ def __init__(self, host, port, id=None, callback=None, max_connections=0):
...
 
             # Cannot connect with yourself
-            if self.id == connected_node_id:
-                print("connect_with_node: You cannot connect with yourself?!")
+            if self.id == connected_node_id or connected_node_id in node_ids:
                 sock.send("CLOSING: Already having a connection together".encode('utf-8'))
                 sock.close()
                 return True
 
-            # Fix bug: Cannot connect with nodes that are already connected with us!
```
Comment (macsnoeren) [inline p2pnetwork/node.py:177]: This bug has not been resolved by the code above. Inbound connections cannot be checked by their port. We should check for a similar node id to check if we are already connected to the node.

**Review by macsnoeren — CHANGES_REQUESTED** [review]
Could you check how to fix the bug as mentioned in the review?

**Comment by macsnoeren** [comment]
@SamuelHaidu Can you help me with this?

**Comment by macsnoeren** [comment]
I have made some changed to development. Most probably you need to fetch development to be able to provide the required changes.

**Comment by SamuelHaidu** [comment]
Sorry maybe you can't see my review :(

Unintentionally I add this verification on the line 155:

`171 if self.id == connected_node_id:`
to:
`155 if self.id == connected_node_id or connected_node_id in node_ids:`

A doubt: Check by ip and port can be removed? Can be seen as redundant

**Review by macsnoeren — APPROVED** [review]
I have made the required changes.

**Pull request merged**